### PR TITLE
Batch

### PR DESCRIPTION
**Pull request by munkk**

### CODE REVIEW

File: core/body/body.ctrl.js [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import { PathService } from '../path';
 import { View } from '../view/view';
 import { Fastdom } from '../services/fastdom';
+import { CellEditor } from '../edit/edit.cell.editor';
+import {get} from '../services/value';
```
Comment (klumba12) [inline core/body/body.ctrl.js]: Fix formatting

File: core/body/body.ctrl.js [hunk 2]
```diff
@@ -177,4 +221,17 @@ export class BodyCtrl extends View {
 	get selection() {
 		return this.model.selection();
 	}
-}
\ No newline at end of file
+
+	isAllowBatch(e, cell) {
+		const tr = cell.element;
```
Comment (klumba12) [inline core/body/body.ctrl.js]: This should be set in the cell-handler

File: core/selection/selection.model.d.ts [hunk 3]
```diff
@@ -37,8 +37,9 @@ export declare class SelectionModel {
 	 * * `'single'`
 	 * * `'multiple'`
 	 * * `'range'`
+	 * * `'batch'`
 	 */
-	mode: 'single' | 'mutiple' | 'range';
+	mode: 'single' | 'mutiple' | 'range' | 'batch';
```
Comment (klumba12) [inline core/selection/selection.model.d.ts]: There shouldnt be mode batch

File: demo/app/home/home.component.html [hunk 4]
```diff
@@ -13,13 +13,14 @@
...
 <mat-card>
-    <q-grid id="q-grid-demo"            
+    <q-grid id="q-grid-demo"
             q-grid-autofocus
             [columns]="columns"
             [rows]="rows"
             [filterFetch]="filterFetch"
             [sortBy]="['+name.last', '-name.first']"
-            editMode="cell"
+            [selectionMode]="'range'"
```
Comment (klumba12) [inline demo/app/home/home.component.html]: Edit mode batch

File: demo/app/home/home.component.html [hunk 5]
```diff
@@ -56,8 +57,8 @@
             </ng-template>
         </q-grid-legend>
 
-        <q-grid-data-manipulation>
-		  </q-grid-data-manipulation>
+        <!--<q-grid-data-manipulation>-->
```
Comment (klumba12) [inline demo/app/home/home.component.html]: Should work with dm plugin

File: core/selection/selection.model.d.ts (hunk 3 above)
Comment (munkk) [reply core/selection/selection.model.d.ts]: image

Without 'batch' it won't work

File: demo/app/home/home.component.html (hunk 5 above)
Comment (munkk) [reply demo/app/home/home.component.html]: no, doesn't work

File: core/body/body.ctrl.js [hunk 6]
```diff
@@ -12,6 +13,7 @@ export class BodyCtrl extends View {
 		this.bag = bag;
 		this.table = table;
 		this.rangeStartCell = null;
+		this.editState = this.model.edit().state;
```
Comment (klumba12) [inline core/body/body.ctrl.js]: remove this one

File: core/body/body.ctrl.js [hunk 7]
```diff
@@ -174,7 +185,42 @@ export class BodyCtrl extends View {
 		}
 	}
 
+	doBatch(startCell) {
```
Comment (klumba12) [inline core/body/body.ctrl.js]: make a separate service under this core/edit/edit.service
and do `new EditService(model, table....).batchUpdate(...`

File: core/body/body.ctrl.js [hunk 8]
```diff
@@ -174,7 +185,42 @@ export class BodyCtrl extends View {
...
+			const label = startCell.label;
+			const value = startCell.value;
+			const initialType = startCell.column.type;
+			const columnIndices = this.model.columnList().index;
```
Comment (klumba12) [inline core/body/body.ctrl.js]: rename to columnIndex

File: core/body/body.ctrl.js [hunk 9]
```diff
@@ -174,7 +185,42 @@ export class BodyCtrl extends View {
 		}
 	}
 
+	doBatch(startCell) {
+			const label = startCell.label;
+			const value = startCell.value;
+			const initialType = startCell.column.type;
```
Comment (klumba12) [inline core/body/body.ctrl.js]: reaname to startColumnType

File: core/body/body.ctrl.js [hunk 10]
```diff
@@ -174,7 +185,42 @@ export class BodyCtrl extends View {
...
+			const columnIndices = this.model.columnList().index;
+			const cells = [];
+
+			this.selection.items.forEach(item => {
```
Comment (klumba12) [inline core/body/body.ctrl.js]: use here and bellow `for(let i = ...` for perfomance

File: core/body/body.ctrl.js [hunk 11]
```diff
@@ -174,7 +185,42 @@ export class BodyCtrl extends View {
...
+				const {row, column} = item;
+				const key = column.key;
+				const columnIndex = columnIndices.indexOf(key);
+				const rowIndex = row.id;
+				const cellView = this.table.body.cell(rowIndex, columnIndex).model();
+
+				cells.push(cellView.model);
+			});
+
+			cells.forEach(cell => {
```
Comment (klumba12) [inline core/body/body.ctrl.js]: get rid of this forEach it can be done in previous

File: core/selection/selection.model.d.ts [hunk 12]
```diff
@@ -37,6 +37,7 @@ export declare class SelectionModel {
 	 * * `'single'`
 	 * * `'multiple'`
 	 * * `'range'`
+	 * * `'batch'`
 	 */
 	mode: 'single' | 'mutiple' | 'range';
```
Comment (klumba12) [inline core/selection/selection.model.d.ts]: add batch here

File: demo/app/home/home.component.html (hunk 4 above)
Comment (klumba12) [reply demo/app/home/home.component.html]: What will be is we remove selectionMode from here?

File: demo/app/home/home.component.html (hunk 5 above)
Comment (klumba12) [reply demo/app/home/home.component.html]: make it work :)

File: src/main/core/cell/cell-handler.component.html [hunk 13]
```diff
@@ -1 +1,5 @@
-<div class="q-grid-marker"></div>
\ No newline at end of file
+<div *ngIf="root.model.edit().mode === 'batch'"
+	  #marker
+	  class="q-grid-marker"
+	  (mousedown)="root.model.edit({state: 'batch'});">
```
Comment (klumba12) [inline src/main/core/cell/cell-handler.component.html]: make (mousedown)="startBatchEdit()"

File: src/theme/material/assets/focus.scss [hunk 14]
```diff
@@ -25,4 +25,16 @@ $duration: 0.15s;
...
+.q-grid-marker {
+	position: absolute;
+	width: 10px;
+	height: 10px;
+	bottom: 0;
+	right: 0;
+	background-color: #C8C8C8;
+	cursor: move;
+}
+
```
Comment (klumba12) [inline src/theme/material/assets/focus.scss]: remove end lines

File: demo/app/home/home.component.html (hunk 4 above)
Comment (munkk) [reply demo/app/home/home.component.html]: then marker will be invisible

File: core/body/body.ctrl.js [hunk 15]
```diff
@@ -64,9 +64,16 @@ export class BodyCtrl extends View {
 			const pathFinder = new PathService(this.bag.body);
 			const cell = pathFinder.cell(e.path);
 
-			const editMode = this.model.edit().mode;
+			const edit = this.model.edit();
+			const editMode = edit.mode;
+			const editMethod = edit.method;
 			if (selectionState.mode === 'range') {
```
Comment (klumba12) [inline core/body/body.ctrl.js]: can we have `selectionState.mode === 'range' || (editMethod === 'batch' && edit.state === 'startBatch')`?

File: core/body/body.ctrl.js [hunk 16]
```diff
@@ -64,9 +64,16 @@ export class BodyCtrl extends View {
...
-				if (!editMode) {
+
+				if(editMethod === 'batch' && edit.state !== 'startBatch') {
+					return;
+				}
+
+				if (editMode || editMethod === 'batch') {
```
Comment (klumba12) [inline core/body/body.ctrl.js]: no sense to have it

File: demo/app/home/home.component.html [hunk 17]
```diff
@@ -13,15 +13,17 @@
...
-	        [sortBy]="['+name.last', '-name.first']"
-	        editMode="cell"
-	        header="NG2 QGRID SANDBOX">
+    <q-grid id="q-grid-demo"
+            q-grid-autofocus
+            [columns]="columns"
+            [rows]="rows"
+            [filterFetch]="filterFetch"
+            [sortBy]="['+name.last', '-name.first']"
+            [selectionMode]="'range'"
```
Comment (klumba12) [inline demo/app/home/home.component.html]: remove this one

File: demo/app/home/home.component.html [hunk 18]
```diff
@@ -13,15 +13,17 @@
...
+            [editMethod]="'batch'"
```
Comment (klumba12) [inline demo/app/home/home.component.html]: `editMethod="batch"` should work too

File: src/main/core/cell/cell-handler.component.html [hunk 19]
```diff
@@ -1 +1,6 @@
-<div class="q-grid-marker"></div>
\ No newline at end of file
+<i *ngIf="markerVisibility"
```
Comment (klumba12) [inline src/main/core/cell/cell-handler.component.html]: rename to `isMarkerVisible`

File: src/main/core/cell/cell-handler.component.ts [hunk 20]
```diff
@@ -72,4 +78,51 @@ export class CellHandlerComponent implements OnInit {
...
+			}
+		});
+
+		model.navigationChanged.watch(e => {
+			if (e.hasChanges('cell')) {
+				const currentCell = e.state.cell;
+
+				if (model.edit().method === 'batch') {
+					if (previousCell) {
+						previousCell.removeChild(this.marker.nativeElement);
```
Comment (klumba12) [inline src/main/core/cell/cell-handler.component.ts]: use `Fastdom.mutate(() => previous... )` for this

File: src/main/core/cell/cell-handler.component.ts [hunk 21]
```diff
@@ -72,4 +78,51 @@ export class CellHandlerComponent implements OnInit {
...
+					}
+
+					const element = currentCell.model.element;
+
+					element.appendChild(this.marker.nativeElement);
```
Comment (klumba12) [inline src/main/core/cell/cell-handler.component.ts]: `Fastdom.mutate` here too

**Pull request merged**